File: product-of-array-except-self/product-of-array-except-self.cs
public class Solution {
    public int[] ProductExceptSelf(int[] nums) {
        
        int length = nums.Length;
        int[] answer = new int[length];
        
        answer[0]=1;
        
        for(int i=1;i<length;i++)
        {
            answer[i] = nums[i-1] * answer[i-1];
        }
        
        int r=1;
        for(int j=length-1;j>=0;j--)
        {
            answer[j] = answer[j] *r;
            r*= nums[j];
        }
        
        return answer;
    }
}

File: array1problems/src/ProductofArrayExceptSelf.java
class ProductofArrayExceptSelf {
    public int[] productExceptSelf(int[] nums) {
        if(nums == null || nums.length == 0){
            return new int[0];
        }
        int[] output = new int[nums.length];
        int productleft=1;

        for(int x=0;x<nums.length;x++){
            output[x]=productleft;
            productleft*=nums[x];
        }

        int productright=1;
        for(int x=nums.length-1;x>=0;x--){
            output[x]*=productright;
            productright*=nums[x];
        }
        return output;
    }
}

//Time complexity = o(n) 
//space complexity =o(1) 